### PR TITLE
PP-9328: Create recurring token in sandbox provider and set payment instrument if flag set during authorisation

### PR DESCRIPTION
**Pull request by gidsg**
Co-authored-by: Alex Bishop <alexbishop1@users.noreply.github.com>

## WHAT YOU DID
Create recurring token in sandbox provider and set payment instrument if flag set during authorisation

## How to test

Tests should make sense

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/charge/service/ChargeService.java [hunk 1]
```diff
@@ -807,34 +819,6 @@ public int count3dsRequiredEvents(String externalId) {
         return chargeDao.count3dsRequiredEventsForChargeExternalId(externalId);
     }
 
-    private CardDetailsEntity buildCardDetailsEntity(AuthCardDetails authCardDetails) {
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/charge/service/ChargeService.java:810]: This has moved to the new `AuthCardDetailsToCardDetailsEntityConverter` class.

Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/charge/service/ChargeService.java:810]: This has moved to the new `AuthCardDetailsToCardDetailsEntityConverter` class.

File: src/main/java/uk/gov/pay/connector/charge/util/AuthCardDetailsToCardDetailsEntityConverter.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+public class AuthCardDetailsToCardDetailsEntityConverter {
+
+    private final NorthAmericanRegionMapper northAmericanRegionMapper;
+
+    @Inject
+    public AuthCardDetailsToCardDetailsEntityConverter(NorthAmericanRegionMapper northAmericanRegionMapper) {
+        this.northAmericanRegionMapper = northAmericanRegionMapper;
+    }
+
+    public CardDetailsEntity convert(AuthCardDetails authCardDetails) {
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/charge/util/AuthCardDetailsToCardDetailsEntityConverter.java:26]: This used to be `ChargeService.buildCardDetailsEntity(AuthCardDetails)`.

Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/charge/util/AuthCardDetailsToCardDetailsEntityConverter.java:26]: This used to be `ChargeService.buildCardDetailsEntity(AuthCardDetails)`.

File: src/main/java/uk/gov/pay/connector/wallets/WalletAuthorisationDataToAuthCardDetailsConverter.java [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+import uk.gov.pay.connector.wallets.model.WalletAuthorisationData;
+import uk.gov.service.payments.commons.model.CardExpiryDate;
+
+import java.time.format.DateTimeFormatter;
+
+public class WalletAuthorisationDataToAuthCardDetailsConverter {
+
+    private static final DateTimeFormatter EXPIRY_DATE_FORMAT = DateTimeFormatter.ofPattern("MM/yy");
+
+    public AuthCardDetails convert(WalletAuthorisationData walletAuthorisationData) {
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/wallets/WalletAuthorisationDataToAuthCardDetailsConverter.java:13]: This used to be `WalletAuthoriseService.authCardDetailsFor(WalletAuthorisationData)`.

Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/wallets/WalletAuthorisationDataToAuthCardDetailsConverter.java:13]: This used to be `WalletAuthoriseService.authCardDetailsFor(WalletAuthorisationData)`.

File: src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java [hunk 4]
```diff
@@ -169,17 +169,6 @@ private void processGatewayAuthorisationResponse(
         return getPaymentProviderFor(chargeEntity).authoriseWallet(authorisationGatewayRequest);
     }
 
-    private AuthCardDetails authCardDetailsFor(WalletAuthorisationData walletAuthorisationData) {
```
Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java:172]: This is now `WalletAuthorisationDataToAuthCardDetailsConverter`.

Comment (alexbishop1) [inline src/main/java/uk/gov/pay/connector/wallets/WalletAuthoriseService.java:172]: This is now `WalletAuthorisationDataToAuthCardDetailsConverter`.

File: src/test/java/uk/gov/pay/connector/charge/service/ChargeServiceTest.java [hunk 5]
```diff
@@ -466,97 +462,7 @@ public void shouldOfferPaymentStateTransition() {
 
         verify(mockTaskQueueService).offerTasksOnStateTransition(chargeSpy);
     }
-
-    @Test
-    public void updateChargeAndEmitEventPostAuthorisation_shouldEmitEvent() {
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/charge/service/ChargeServiceTest.java:471]: Movedi into `ChargeServicePostAuthorisationTest`.

Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/charge/service/ChargeServiceTest.java:471]: Movedi into `ChargeServicePostAuthorisationTest`.

File: src/test/java/uk/gov/pay/connector/charge/service/ChargeServiceTest.java [hunk 6]
```diff
@@ -466,97 +462,7 @@ public void shouldOfferPaymentStateTransition() {
...
-        AuthCardDetails authCardDetails = new AuthCardDetails();
-        authCardDetails.setCardNo("1234567890");
-        service.updateChargeAndEmitEventPostAuthorisation(chargeSpy.getExternalId(), ENTERING_CARD_DETAILS,
-                authCardDetails, null, null, null, null, null);
-
-        verify(mockEventService).emitAndRecordEvent(PaymentDetailsEntered.from(chargeSpy));
-    }
-
-    @Test
-    public void updateChargeAndEmitEventPostAuthorisation_shouldEmitEventWhenCountryIsInUnitedStates() {
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/charge/service/ChargeServiceTest.java:490]: Now covered by tests in `ChargeServicePostAuthorisationTest` and `AuthCardDetailsToCardDetailsEntityConverterTest`.

Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/charge/service/ChargeServiceTest.java:490]: Now covered by tests in `ChargeServicePostAuthorisationTest` and `AuthCardDetailsToCardDetailsEntityConverterTest`.

File: src/test/java/uk/gov/pay/connector/charge/service/ChargeServiceTest.java [hunk 7]
```diff
@@ -466,97 +462,7 @@ public void shouldOfferPaymentStateTransition() {
...
-        when(chargeSpy.getEvents()).thenReturn(List.of(chargeEvent));
-
-        service.updateChargeAndEmitEventPostAuthorisation(chargeSpy.getExternalId(), ENTERING_CARD_DETAILS,
-                authCardDetails, null, null, null, null, null);
-
-        verify(mockEventService).emitAndRecordEvent(PaymentDetailsEntered.from(chargeSpy));
-    }
-
-    @Test
-    public void shouldUpdateStateOrProvinceForChargeWhenCountryIsUs() {
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/charge/service/ChargeServiceTest.java:514]: Now covered by tests in `ChargeServicePostAuthorisationTest` and `AuthCardDetailsToCardDetailsEntityConverterTest`.

Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/charge/service/ChargeServiceTest.java:514]: Now covered by tests in `ChargeServicePostAuthorisationTest` and `AuthCardDetailsToCardDetailsEntityConverterTest`.

File: src/test/java/uk/gov/pay/connector/charge/service/ChargeServiceTest.java [hunk 8]
```diff
@@ -466,97 +462,7 @@ public void shouldOfferPaymentStateTransition() {
...
-        when(chargeSpy.getEvents()).thenReturn(List.of(chargeEvent));
-
-        service.updateChargeAndEmitEventPostAuthorisation(chargeSpy.getExternalId(), ENTERING_CARD_DETAILS,
-                authCardDetails, null, null, null, null, null);
-
-        assertThat(chargeSpy.getCardDetails().getBillingAddress().get().getStateOrProvince(), is("DC"));
-    }
-
-    @Test
-    public void shouldNotUpdateStateOrProvinceForChargeWhenCountryIsNotInNorthAmerica() {
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/charge/service/ChargeServiceTest.java:538]: Now covered by tests in `ChargeServicePostAuthorisationTest` and `AuthCardDetailsToCardDetailsEntityConverterTest`.

Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/charge/service/ChargeServiceTest.java:538]: Now covered by tests in `ChargeServicePostAuthorisationTest` and `AuthCardDetailsToCardDetailsEntityConverterTest`.

File: src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardAuthoriseServiceTest.java [hunk 9]
```diff
@@ -217,25 +216,8 @@ public void doAuthoriseWithNonCorporateCard_shouldRespondAuthorisationSuccess()
...
-        assertThat(charge.getStatus(), is(AUTHORISATION_SUCCESS.getValue()));
-        assertThat(charge.getGatewayTransactionId(), is(TRANSACTION_ID));
-        verify(mockedChargeEventDao).persistChargeEventOf(eq(charge), isNull());
-        assertThat(charge.get3dsRequiredDetails(), is(nullValue()));
-        assertThat(charge.getCardDetails(), is(notNullValue()));
-        assertThat(charge.getCorporateSurcharge().isPresent(), is(false));
-    }
-
-    @Test
-    public void doAuthoriseWithNonCorporateCard_shouldRespondAuthorisationSuccess_2() throws Exception {
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardAuthoriseServiceTest.java:235]: Duplicate test.

Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardAuthoriseServiceTest.java:235]: Duplicate test.

File: src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardAuthoriseServiceTest.java [hunk 10]
```diff
@@ -246,30 +228,7 @@ public void doAuthoriseWithNonCorporateCard_shouldRespondAuthorisationSuccess_2(
         assertThat(charge.getGatewayTransactionId(), is(TRANSACTION_ID));
         verify(mockedChargeEventDao).persistChargeEventOf(eq(charge), isNull());
         assertThat(charge.get3dsRequiredDetails(), is(nullValue()));
-        assertThat(charge.getCardDetails(), is(notNullValue()));
-        assertThat(charge.getCorporateSurcharge().isPresent(), is(false));
-    }
-
-    @Test
-    public void doAuthoriseWithoutBillingAddress_shouldRespondAuthorisationSuccess() throws Exception {
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardAuthoriseServiceTest.java:254]: Moved to `AuthCardDetailsToCardDetailsEntityConverterTest`.

Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardAuthoriseServiceTest.java:254]: Moved to `AuthCardDetailsToCardDetailsEntityConverterTest`.

File: src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardAuthoriseServiceTest.java [hunk 11]
```diff
@@ -494,66 +459,14 @@ public void shouldRespondAuthorisationError() throws Exception {
         assertThat(charge.getGatewayTransactionId(), is(nullValue()));
     }
 
-    @Test
-    public void doAuthorise_shouldStoreExpectedCardDetails_whenAuthorisationSuccess() throws Exception {
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardAuthoriseServiceTest.java:498]: Moved to `AuthCardDetailsToCardDetailsEntityConverterTest`.

Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/paymentprocessor/service/CardAuthoriseServiceTest.java:498]: Moved to `AuthCardDetailsToCardDetailsEntityConverterTest`.

File: src/test/java/uk/gov/pay/connector/wallets/WalletAuthoriseServiceTest.java [hunk 12]
```diff
@@ -256,41 +273,6 @@ public void doAuthoriseCard_GooglePay_shouldRespondAuthorisationSuccess() throws
         assertThat(eventCaptor.getValue().getEventType(), is("PAYMENT_DETAILS_ENTERED"));
     }
 
-    @Test
-    public void doAuthoriseCard_ApplePay_shouldRespondAuthorisationSuccess_whenCardNumberIsEmptyString() throws Exception {
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/wallets/WalletAuthoriseServiceTest.java:260]: Moved to `WalletAuthorisationDataToAuthCardDetailsConverterTest`.

Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/wallets/WalletAuthoriseServiceTest.java:260]: Moved to `WalletAuthorisationDataToAuthCardDetailsConverterTest`.

File: src/test/java/uk/gov/pay/connector/wallets/WalletAuthoriseServiceTest.java [hunk 13]
```diff
@@ -363,30 +345,14 @@ public void shouldRespondAuthorisationError() throws Exception {
         assertThat(charge.getWalletType(), is(WalletType.APPLE_PAY));
     }
 
-    @Test
-    public void doAuthorise_shouldStoreExpectedCardDetails_whenAuthorisationSuccess() throws Exception {
```
Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/wallets/WalletAuthoriseServiceTest.java:367]: Moved to `WalletAuthorisationDataToAuthCardDetailsConverterTest`.

Comment (alexbishop1) [inline src/test/java/uk/gov/pay/connector/wallets/WalletAuthoriseServiceTest.java:367]: Moved to `WalletAuthorisationDataToAuthCardDetailsConverterTest`.

File: src/main/java/uk/gov/pay/connector/charge/util/AuthCardDetailsToCardDetailsEntityConverter.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+import javax.inject.Inject;
+
+import static uk.gov.pay.connector.common.model.domain.NumbersInStringsSanitizer.sanitize;
+
+public class AuthCardDetailsToCardDetailsEntityConverter {
+
+    private final NorthAmericanRegionMapper northAmericanRegionMapper;
+
+    @Inject
+    public AuthCardDetailsToCardDetailsEntityConverter(NorthAmericanRegionMapper northAmericanRegionMapper) {
```
Comment (sfount) [inline src/main/java/uk/gov/pay/connector/charge/util/AuthCardDetailsToCardDetailsEntityConverter.java:22]: 👍  having this in a easy to test, shareable place looks looks particularly useful as different APIs will be considering authorisation from different angles

File: src/main/java/uk/gov/pay/connector/gateway/sandbox/SandboxPaymentProvider.java [hunk 15]
```diff
@@ -44,6 +51,8 @@ public class SandboxPaymentProvider implements PaymentProvider, SandboxGatewayRe
 
     private SandboxWalletAuthorisationHandler sandboxWalletAuthorisationHandler;
 
+    
+    @Inject
```
Comment (sfount) [inline src/main/java/uk/gov/pay/connector/gateway/sandbox/SandboxPaymentProvider.java]: I'm guessing this injection hook is no longer needed

File: src/main/java/uk/gov/pay/connector/paymentinstrument/service/PaymentInstrumentService.java [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    @Transactional
+    public PaymentInstrumentEntity createPaymentInstrument(CardDetailsEntity cardDetails, Map<String, String> recurringAuthToken) {
+        var now = clock.instant();
+        var paymentInstrument = new PaymentInstrumentEntity.PaymentInstrumentEntityBuilder()
+                .withCreatedDate(now)
+                .withRecurringAuthToken(recurringAuthToken)
+                .withCardDetails(cardDetails)
+                .withStatus(PaymentInstrumentStatus.ACTIVE)
```
Comment (sfount) [inline src/main/java/uk/gov/pay/connector/paymentinstrument/service/PaymentInstrumentService.java]: I think the default payment instrument state should be `CREATED`. 

The proposed flow in the spec is

`CREATED` - payment is attempted/ authorised - payment instrument set up before/ after authorisation depending on provider integration. This fits this initial created state
`ACTIVE` - payment is guaranteed queued for capture - user has been made aware of details being saved and has pressed "confirm". We know the tokens stored are good to be used from this point

File: src/main/java/uk/gov/pay/connector/wallets/WalletAuthorisationDataToAuthCardDetailsConverter.java (hunk 3 above)
Comment (sfount) [inline src/main/java/uk/gov/pay/connector/wallets/WalletAuthorisationDataToAuthCardDetailsConverter.java:13]: As each of these is a relatively simple interface without too many dependencies I wonder if having one card details converter that can convert for wallets and web card details would keep things tidy in one place. I think this pattern is absolutely fine and don't suggest it should be changed for this pr

**Review by sfount — COMMENTED** [review]
Really clear code and what feels like a helpful refactor to make some quite branching code easier to test 👍  one comment on the initial state of payment instruments but otherwise looks good to go

File: src/main/java/uk/gov/pay/connector/gateway/sandbox/SandboxPaymentProvider.java (hunk 15 above)
Comment (alexbishop1) [reply src/main/java/uk/gov/pay/connector/gateway/sandbox/SandboxPaymentProvider.java]: Removed.

File: src/main/java/uk/gov/pay/connector/paymentinstrument/service/PaymentInstrumentService.java (hunk 16 above)
Comment (alexbishop1) [reply src/main/java/uk/gov/pay/connector/paymentinstrument/service/PaymentInstrumentService.java]: Changed.

**Review by sfount — APPROVED** [review]
Follow up changes look good, this feels like a very neat pattern for the next providers 👍

**Pull request merged**